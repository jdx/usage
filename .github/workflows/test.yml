on:
  push:
    branches: ['main']
  pull_request:

concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true

env:
  MISE_EXPERIMENTAL: 1

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: jdx/mise-action@v2
      - run: mise r ci
