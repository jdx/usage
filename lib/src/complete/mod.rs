use crate::Spec;

mod bash;
mod fish;
mod zsh;

pub struct CompleteOptions {
    pub shell: String,
    pub bin: String,
    pub cache_key: Option<String>,
    pub spec: Option<Spec>,
    pub usage_cmd: Option<String>,
}

pub fn complete(options: &CompleteOptions) -> String {
    match options.shell.as_str() {
        "bash" => bash::complete_bash(options),
        "fish" => fish::complete_fish(options),
        "zsh" => zsh::complete_zsh(options),
        _ => unimplemented!("unsupported shell: {}", options.shell),
    }
}
