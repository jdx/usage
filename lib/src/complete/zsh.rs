// use crate::env;

pub fn complete_zsh(bin: &str, usage_cmd: &str) -> String {
    // let usage = env::USAGE_BIN.display();
    // let cmds = vec![&spec.cmd];
    // let args = render_args(&cmds);
    format!(
        r#"
#compdef {bin}
local curcontext="$curcontext"

# caching config
_usage_{bin}_cache_policy() {{
  if [[ -z "${{lifetime}}" ]]; then
    lifetime=$((60*60*4)) # 4 hours
  fi
  local -a oldp
  oldp=( "$1"(Nms+${{lifetime}}) )
  (( $#oldp ))
}}

_{bin}() {{
  typeset -A opt_args
  local curcontext="$curcontext" spec cache_policy

  if ! command -v usage &> /dev/null; then
      echo >&2
      echo "Error: usage CLI not found. This is required for completions to work in {bin}." >&2
      echo "See https://usage.jdx.dev for more information." >&2
      return 1
  fi

  zstyle -s ":completion:${{curcontext}}:" cache-policy cache_policy
  if [[ -z $cache_policy ]]; then
    zstyle ":completion:${{curcontext}}:" cache-policy _usage_{bin}_cache_policy
  fi

  if ( [[ -z "${{_usage_{bin}_spec:-}}" ]] || _cache_invalid _usage_{bin}_spec ) \
      && ! _retrieve_cache _usage_{bin}_spec;
  then
    spec="$({usage_cmd})"
    _store_cache _usage_{bin}_spec spec
  fi

  _arguments "*: :(($(usage complete-word --shell zsh -s "$spec" -- "${{words[@]}}" )))"
  return 0
}}

if [ "$funcstack[1]" = "_{bin}" ]; then
    _{bin} "$@"
else
    compdef _{bin} {bin}
fi

# vim: noet ci pi sts=0 sw=4 ts=4
"#
    )
}

// fn render_args(cmds: &[&SchemaCmd]) -> String {
//     format!("XX")
// }

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;

    #[test]
    fn test_complete_zsh() {
        // let spec = r#"
        // "#;
        // let spec = Spec::parse(&Default::default(), spec).unwrap();
        assert_snapshot!(complete_zsh("mycli", "mycli complete --usage").trim(), @r###"
        #compdef mycli
        local curcontext="$curcontext"

        # caching config
        _usage_mycli_cache_policy() {
          if [[ -z "${lifetime}" ]]; then
            lifetime=$((60*60*4)) # 4 hours
          fi
          local -a oldp
          oldp=( "$1"(Nms+${lifetime}) )
          (( $#oldp ))
        }

        _mycli() {
          typeset -A opt_args
          local curcontext="$curcontext" spec cache_policy

          if ! command -v usage &> /dev/null; then
              echo >&2
              echo "Error: usage CLI not found. This is required for completions to work in mycli." >&2
              echo "See https://usage.jdx.dev for more information." >&2
              return 1
          fi

          zstyle -s ":completion:${curcontext}:" cache-policy cache_policy
          if [[ -z $cache_policy ]]; then
            zstyle ":completion:${curcontext}:" cache-policy _usage_mycli_cache_policy
          fi

          if ( [[ -z "${_usage_mycli_spec:-}" ]] || _cache_invalid _usage_mycli_spec ) \
              && ! _retrieve_cache _usage_mycli_spec;
          then
            spec="$(mycli complete --usage)"
            _store_cache _usage_mycli_spec spec
          fi

          _arguments "*: :(($(usage complete-word --shell zsh -s "$spec" -- "${words[@]}" )))"
          return 0
        }

        if [ "$funcstack[1]" = "_mycli" ]; then
            _mycli "$@"
        else
            compdef _mycli mycli
        fi

        # vim: noet ci pi sts=0 sw=4 ts=4
        "###);
    }
}
