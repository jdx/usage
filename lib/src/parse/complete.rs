use serde::{Deserialize, Serialize};

use crate::error::UsageErr;
use crate::parse::context::ParsingContext;
use crate::parse::helpers::NodeHelper;

#[derive(Debug, Default, Clone, Serialize, Deserialize)]
pub struct Complete {
    pub name: String,
    pub run: Option<String>,
    pub type_: Option<String>,
}

impl Complete {
    pub(crate) fn parse(ctx: &ParsingContext, node: &NodeHelper) -> Result<Self, UsageErr> {
        let mut config = Self::default();
        node.ensure_arg_len(1..=1)?;
        config.name = node.arg(0)?.ensure_string()?.to_string().to_lowercase();
        for (k, v) in node.props() {
            match k {
                "run" => {
                    if config.type_.is_some() {
                        bail_parse!(ctx, *v.entry.span(), "can set run or type, not both")
                    }
                    config.run = Some(v.ensure_string()?.to_string())
                }
                "type" => {
                    if config.run.is_some() {
                        bail_parse!(ctx, *v.entry.span(), "can set run or type, not both")
                    }
                    config.type_ = Some(v.ensure_string()?.to_string())
                }
                k => bail_parse!(ctx, *v.entry.span(), "unsupported complete key {k}"),
            }
        }
        Ok(config)
    }
}
