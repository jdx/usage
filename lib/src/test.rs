use crate::Spec;
use once_cell::sync::Lazy;

#[macro_export]
macro_rules! spec {
    { $spec:literal } => {
        $spec.parse::<$crate::spec::Spec>()
    };
}

pub static SPEC_KITCHEN_SINK: Lazy<Spec> = Lazy::new(|| {
    spec! {r#"
bin "mycli"
arg "arg1" help="arg1 description"
arg "arg2" help="arg2 description" default="default value" {
    choices "choice1" "choice2" "choice3"
}
arg "arg3" help="arg3 description" required=true long_help="arg3 long description"
arg "argrest" var=true

flag "--flag1" help="flag1 description"
flag "--flag2" help="flag2 description" long_help="flag2 long description"
flag "--flag3" help="flag3 description" negate="--no-flag3"

flag "--shell <shell>" {
  choices "bash" "zsh" "fish"
}

cmd "plugin" {
  cmd "install" {
    arg "plugin"
    arg "version"
    flag "-g --global"
    flag "-d --dir <dir>"
    flag "-f --force" negate="--no-force"
  }
}

complete "plugin" run="echo \"plugin-1\nplugin-2\nplugin-3\""
"#}
    .unwrap()
});

#[test]
fn test_parse() {
    assert_eq!(SPEC_KITCHEN_SINK.name, "mycli");
}
