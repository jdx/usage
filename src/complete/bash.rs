use crate::{env, Spec};

pub fn complete_bash(spec: &Spec) -> String {
    let usage = &*env::USAGE_CMD;
    let bin = &spec.bin;
    let raw = shell_escape::unix::escape(spec.to_string().into());
    format!(
        r#"
_{bin}() {{
    local raw
    spec={raw}

    COMPREPLY=($({usage} complete-word -s "$spec" --cword="$COMP_CWORD" -- "${{COMP_WORDS[@]}}"))
    #COMPREPLY=($(compgen -W "${{COMPREPLY[*]}}" -- "${{COMP_WORDS[$COMP_CWORD]}}"))
    return 0
}}

complete -F _{bin} {bin}
# vim: noet ci pi sts=0 sw=4 ts=4 ft=sh
"#
    )
}

// fn render_args(cmds: &[&SchemaCmd]) -> String {
//     format!("XX")
// }

#[cfg(test)]
mod tests {
    use crate::parse::spec::Spec;

    use super::*;

    #[test]
    fn test_complete_bash() {
        let spec = r#"
        "#;
        let spec = Spec::parse(&Default::default(), spec).unwrap();
        assert_snapshot!(complete_bash(&spec).trim(), @r###"
        _() {
            local raw
            spec=''

            COMPREPLY=($(usage complete-word -s "$spec" --cword="$COMP_CWORD" -- "${COMP_WORDS[@]}"))
            #COMPREPLY=($(compgen -W "${COMPREPLY[*]}" -- "${COMP_WORDS[$COMP_CWORD]}"))
            return 0
        }

        complete -F _ 
        # vim: noet ci pi sts=0 sw=4 ts=4 ft=sh
        "###);
    }
}
