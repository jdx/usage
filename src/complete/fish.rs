use crate::{env, Spec};

pub fn complete_fish(spec: &Spec) -> String {
    let usage = &*env::USAGE_CMD;
    let bin = &spec.bin;
    let raw = spec.to_string().replace('\'', r"\'").to_string();
    format!(
        r#"
set _usage_spec_{bin} '{raw}'
complete -xc {bin} -a '({usage} complete-word -s "$_usage_spec_{bin}" --ctoken=(commandline -t) -- (commandline -op))'
"#
    )
}

#[cfg(test)]
mod tests {
    use crate::parse::spec::Spec;

    use super::*;

    #[test]
    fn test_complete_fish() {
        let spec = r#"
        "#;
        let spec = Spec::parse(&Default::default(), spec).unwrap();
        assert_snapshot!(complete_fish(&spec).trim(), @r###"
        set _usage_spec_ ''
        complete -xc  -a '(usage complete-word -s "$_usage_spec_" --ctoken=(commandline -t) -- (commandline -op))'
        "###);
    }
}
