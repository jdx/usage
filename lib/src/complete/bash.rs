use crate::complete::CompleteOptions;
use heck::ToSnakeCase;

pub fn complete_bash(opts: &CompleteOptions) -> String {
    let bin = &opts.bin;
    let bin_snake = bin.to_snake_case();
    let spec_variable = if let Some(cache_key) = &opts.cache_key {
        format!("_usage_spec_{bin_snake}_{}", cache_key.to_snake_case())
    } else {
        format!("_usage_spec_{bin_snake}")
    };
    let mut out = vec![format!(
        r#"_{bin_snake}() {{
    if ! command -v usage &> /dev/null; then
        echo >&2
        echo "Error: usage CLI not found. This is required for completions to work in {bin}." >&2
        echo "See https://usage.jdx.dev for more information." >&2
        return 1
    fi"#
    )];

    if let Some(usage_cmd) = &opts.usage_cmd {
        out.push(format!(
            r#"
    if [[ -z ${{{spec_variable}:-}} ]]; then
        {spec_variable}="$({usage_cmd})"
    fi"#
        ));
    }

    if let Some(spec) = &opts.spec {
        out.push(format!(
            r#"
    read -r -d '' {spec_variable} <<'__USAGE_EOF__'
{spec}
__USAGE_EOF__"#,
            spec = spec.to_string().trim()
        ));
    }

    out.push(format!(
        r#"
    # shellcheck disable=SC2207
    COMPREPLY=( $(usage complete-word --shell bash -s "${{{spec_variable}}}" --cword="$COMP_CWORD" -- "${{COMP_WORDS[@]}}" ) )
    # shellcheck disable=SC2181
    if [[ $? -ne 0 ]]; then
        unset COMPREPLY
    fi
    return 0
}}

shopt -u hostcomplete && complete -o nospace -o bashdefault -o nosort -F _{bin_snake} {bin}
# vim: noet ci pi sts=0 sw=4 ts=4 ft=sh
"#
    ));

    out.join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test::SPEC_KITCHEN_SINK;
    use insta::assert_snapshot;

    #[test]
    fn test_complete_bash() {
        assert_snapshot!(complete_bash(&CompleteOptions {
            shell: "bash".to_string(),
            bin: "mycli".to_string(),
            cache_key: None,
            spec: None,
            usage_cmd: Some("mycli complete --usage".to_string()),
        }));
        assert_snapshot!(complete_bash(&CompleteOptions {
            shell: "bash".to_string(),
            bin: "mycli".to_string(),
            cache_key: Some("1.2.3".to_string()),
            spec: None,
            usage_cmd: Some("mycli complete --usage".to_string()),
        }));
        assert_snapshot!(complete_bash(&CompleteOptions {
            shell: "bash".to_string(),
            bin: "mycli".to_string(),
            cache_key: None,
            spec: Some(SPEC_KITCHEN_SINK.clone()),
            usage_cmd: None,
        }));
    }
}
