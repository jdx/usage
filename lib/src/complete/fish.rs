use crate::Spec;

pub fn complete_fish(bin: &str, usage_cmd: Option<&str>, spec: Option<&Spec>) -> String {
    let mut out = vec![format!(
        r#"
# if "usage" is not installed show an error
if ! command -v usage &> /dev/null
    echo >&2
    echo "Error: usage CLI not found. This is required for completions to work in {bin}." >&2
    echo "See https://usage.jdx.dev for more information." >&2
    return 1
end"#
    )];

    if let Some(usage_cmd) = &usage_cmd {
        out.push(format!(
            r#"
set _usage_spec_{bin} ({usage_cmd} | string collect)"#
        ));
    }

    if let Some(spec) = &spec {
        let spec_escaped = spec.to_string().replace("'", r"\'");
        out.push(format!(
            r#"
set -x _usage_spec_{bin} '{spec_escaped}'"#
        ));
    }

    out.push(format!(
        r#"complete -xc {bin} -a '(usage complete-word --shell fish -s "$_usage_spec_{bin}" -- (commandline -cop) (commandline -t))'"#
    ));

    out.join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test::SPEC_KITCHEN_SINK;
    use insta::assert_snapshot;

    #[test]
    fn test_complete_fish() {
        // let spec = r#"
        // "#;
        // let spec = Spec::parse(&Default::default(), spec).unwrap();
        assert_snapshot!(complete_fish("mycli", Some("mycli complete --usage"), None).trim(), @r###"
        # if "usage" is not installed show an error
        if ! command -v usage &> /dev/null
            echo >&2
            echo "Error: usage CLI not found. This is required for completions to work in mycli." >&2
            echo "See https://usage.jdx.dev for more information." >&2
            return 1
        end

        set _usage_spec_mycli (mycli complete --usage | string collect)
        complete -xc mycli -a '(usage complete-word --shell fish -s "$_usage_spec_mycli" -- (commandline -cop) (commandline -t))'
        "###);

        assert_snapshot!(complete_fish("mycli", None, Some(&SPEC_KITCHEN_SINK)).trim(), @r##"
        # if "usage" is not installed show an error
        if ! command -v usage &> /dev/null
            echo >&2
            echo "Error: usage CLI not found. This is required for completions to work in mycli." >&2
            echo "See https://usage.jdx.dev for more information." >&2
            return 1
        end

        set -x _usage_spec_mycli 'name "mycli"
        bin "mycli"
        source_code_link_template "https://github.com/jdx/mise/blob/main/src/cli/{{path}}.rs"
        flag "--flag1" help="flag1 description"
        flag "--flag2" help="flag2 description" {
            long_help "flag2 long description"
        }
        flag "--flag3" help="flag3 description" negate="--no-flag3"
        flag "--shell" {
            arg "<shell>" {
                choices "bash" "zsh" "fish"
            }
        }
        arg "<arg1>" help="arg1 description"
        arg "<arg2>" help="arg2 description" default="default value" {
            choices "choice1" "choice2" "choice3"
        }
        arg "<arg3>" help="arg3 description" help_long="arg3 long description"
        arg "<argrest>..." var=true
        cmd "plugin" {
            cmd "install" {
                flag "-g --global"
                flag "-d --dir" {
                    arg "<dir>"
                }
                flag "-f --force" negate="--no-force"
                arg "<plugin>"
                arg "<version>"
            }
        }
        '
        complete -xc mycli -a '(usage complete-word --shell fish -s "$_usage_spec_mycli" -- (commandline -cop) (commandline -t))'
        "##);
    }
}
