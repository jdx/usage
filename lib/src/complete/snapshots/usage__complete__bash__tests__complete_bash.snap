---
source: lib/src/complete/bash.rs
expression: "complete_bash(&CompleteOptions\n{\n    shell: \"bash\".to_string(), bin: \"mycli\".to_string(), cache_key: None,\n    spec: None, usage_cmd: Some(\"mycli complete --usage\".to_string()),\n})"
snapshot_kind: text
---
_mycli() {
    if ! command -v usage &> /dev/null; then
        echo >&2
        echo "Error: usage CLI not found. This is required for completions to work in mycli." >&2
        echo "See https://usage.jdx.dev for more information." >&2
        return 1
    fi

    if [[ -z ${_usage_spec_mycli:-} ]]; then
        _usage_spec_mycli="$(mycli complete --usage)"
    fi

    # shellcheck disable=SC2207
    COMPREPLY=( $(usage complete-word --shell bash -s "${_usage_spec_mycli}" --cword="$COMP_CWORD" -- "${COMP_WORDS[@]}" ) )
    # shellcheck disable=SC2181
    if [[ $? -ne 0 ]]; then
        unset COMPREPLY
    fi
    return 0
}

shopt -u hostcomplete && complete -o nospace -o bashdefault -o nosort -F _mycli mycli
# vim: noet ci pi sts=0 sw=4 ts=4 ft=sh
